processing/src/main/java/io/druid/segment/filter/Filters.java

```diff
@@ -119,4 +128,170 @@ public Filter apply(@Nullable String input)
 
     return filter;
   }
-}
+
+  public static Filter[] partitionWithBitmapSupport(Filter current)
+  {
+    return partitionFilterWith(
+        current, new Predicate<Filter>()
+        {
+          @Override
+          public boolean apply(Filter input)
+          {
+            return input.supportsBitmap();
+          }
+        }
+    );
+  }
+
+  private static Filter[] partitionFilterWith(Filter current, Predicate<Filter> predicate)
+  {
+    if (current == null) {
+      return null;
+    }
+    current = pushDownNot(current);
+    current = convertToCNF(current);
+
+    List<Filter> bitmapIndexSupported = Lists.newArrayList();
+    List<Filter> bitmapIndexNotSupported = Lists.newArrayList();
+
+    traverse(current, predicate, bitmapIndexSupported, bitmapIndexNotSupported);
+
+    return new Filter[]{andFilter(bitmapIndexSupported), andFilter(bitmapIndexNotSupported)};
+  }
+
+  private static void traverse(
+      Filter current,
+      Predicate<Filter> predicate,
+      List<Filter> support,
+      List<Filter> notSupport
+  )
+  {
+    if (current instanceof AndFilter) {
+      for (Filter child : ((AndFilter) current).getFilters()) {
+        traverse(child, predicate, support, notSupport);
+      }
+    } else {
+      if (predicate.apply(current)) {
+        support.add(current);
+      } else {
+        notSupport.add(current);
+      }
+    }
+  }
+
+  private static Filter andFilter(List<Filter> filters) {
+    return filters.isEmpty() ? null : filters.size() == 1 ? filters.get(0) : new AndFilter(filters);
+  }
+
+  // copied from apache hive
+  private static Filter pushDownNot(Filter current)
```

Comment:
we should mention as part of the NOTICE file that this contains code adapted from Apache Hive.